File: events/bus.go
package events

import (
	"context"
	"log"
	"sync"
)

type EventHandler func(ctx context.Context, event Event) error

type EventBus interface {
	Subscribe(eventType EventType, handler EventHandler)
	Publish(ctx context.Context, event Event)
}

type MemoryEventBus struct {
	mu       sync.RWMutex
	handlers map[EventType][]EventHandler
}

func NewMemoryEventBus() *MemoryEventBus {
	return &MemoryEventBus{
		handlers: make(map[EventType][]EventHandler),
	}
}

func (b *MemoryEventBus) Subscribe(eventType EventType, handler EventHandler) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.handlers[eventType] = append(b.handlers[eventType], handler)
	log.Printf("[EVENT BUS] Subscribed handler to event '%s'", eventType)
}

func (b *MemoryEventBus) Publish(ctx context.Context, event Event) {
	b.mu.RLock()
	handlers, exists := b.handlers[event.GetType()]
	b.mu.RUnlock()

	if !exists || len(handlers) == 0 {
		return
	}

	log.Printf("[EVENT BUS] Publishing Event '%s' for Doctor '%s'", event.GetType(), event.GetDoctorID())

	for _, handler := range handlers {
		// Non-blocking async event dispatch
		go func(h EventHandler) {
			if err := h(ctx, event); err != nil {
				log.Printf("[EVENT BUS] Error handling event '%s': %v", event.GetType(), err)
			}
		}(handler)
	}
}

File: backend_api/events/bus.py
from threading import Lock
import threading

class MemoryEventBus:
    def __init__(self):
        self.mu = Lock()
        self.handlers = {}

    def subscribe(self, event_type: str, handler):
        with self.mu:
            if event_type not in self.handlers:
                self.handlers[event_type] = []
            self.handlers[event_type].append(handler)
            print(f"[EVENT BUS] Subscribed handler to event '{event_type}'")

    def publish(self, context, event):
        with self.mu:
            handlers = self.handlers.get(event.get_type(), [])

        if not handlers:
            return

        print(f"[EVENT BUS] Publishing Event '{event.get_type()}' for Doctor '{event.get_doctor_id()}'")

        for handler in handlers:
            def run_handler(h):
                try:
                    h(context, event)
                except Exception as e:
                    print(f"[EVENT BUS] Error handling event '{event.get_type()}': {e}")
            
            threading.Thread(target=run_handler, args=(handler,)).start()
